### ISSUE TITLE
argmax/argmin wrong for arrays with infinities

### ISSUE THREAD

**Issue by asmeurer** [cupy/cupy#7424]
### Description

```py
>>> cupy.argmax(cupy.array([-cupy.inf]))
array(1)
>>> cupy.argmin(cupy.array([cupy.inf]))
array(1)
```

compare:

```py
>>> np.argmax(np.array([-cupy.inf]))
0
>>> np.argmin(np.array([cupy.inf]))
0
```

### To Reproduce

```py
# Write the code here
```

### Installation

Conda-Forge (`conda install ...`)

### Environment

```
OS                           : Linux-5.13.0-51-generic-x86_64-with-glibc2.34
Python Version               : 3.10.4
CuPy Version                 : 11.5.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.24.1
SciPy Version                : 1.9.3
Cython Build Version         : 0.29.33
Cython Runtime Version       : 0.29.32
CUDA Root                    : /home/aaronmeurer/anaconda3/envs/array-apis
nvcc PATH                    : /home/aaronmeurer/anaconda3/envs/array-apis/bin/nvcc
CUDA Build Version           : 11020
CUDA Driver Version          : 11070
CUDA Runtime Version         : 11080
cuBLAS Version               : (available)
cuFFT Version                : 10900
cuRAND Version               : 10300
cuSOLVER Version             : (11, 4, 1)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 8)
Thrust Version               : 101000
CUB Build Version            : 101000
Jitify Build Version         : b8d229d
cuDNN Build Version          : 8401
cuDNN Version                : 8401
NCCL Build Version           : 21403
NCCL Runtime Version         : 21403
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : NVIDIA GeForce RTX 2080
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:01:00.0
```

**Comment by asmeurer** [comment]
`min` and `max` has a similar issue:

```
>>> cupy.min(cupy.array([cupy.inf]))
array(1.79769313e+308)
>>> cupy.max(cupy.array([-cupy.inf]))
array(-1.79769313e+308)
```

**Comment by asi1024** [comment]
@leofang I think that this issue may be due to a bug in CUB.
The identity of `cub::DeviceReduce::Min` is defined as `Traits<InitT>::Max()` [here](https://github.com/NVIDIA/cub/blob/5d12837f92ee12016827ad6f1ccbbc963eb428ff/cub/device/device_reduce.cuh#L494). It is defined as `FLT_MAX` [here](https://github.com/NVIDIA/cub/blob/5d12837f92ee12016827ad6f1ccbbc963eb428ff/cub/util_type.cuh#L1090) when the data type is float32, but It should be `FLT_INF`.
Now I am currently working on a minimal reproducer and will post an issue at https://github.com/NVIDIA/cub! :smiley:

**Comment by kmaehashi** [comment]
In case anyone is in need an immediate workaround, please set `CUPY_ACCELERATORS=""` environment variable.

**Comment by leofang** [comment]
@asi1024 I am not sure if this is the root cause. The quoted lines basically are a homegrown `std::numeric_limits<T>::{max,lowest}`, and indeed in C++ they are defined as `FLT_MAX` and `-FLT_MAX`, respectively, not `FLT_INF`/`-FLT_INF`. I didn't check but my feeling is the bug is caused by something else.

**Comment by kmaehashi** [comment]
I saw @asi1024 just upstream this to https://github.com/NVIDIA/cub/issues/642.

> @asi1024 I am not sure if this is the root cause. The quoted lines basically are CUB's homegrown implementation for `std::numeric_limits<T>::{max,lowest}`, and indeed in C++ they are defined as `FLT_MAX` and `-FLT_MAX`, respectively, not `FLT_INF`/`-FLT_INF`. I didn't check but my feeling is the bug is caused by something else.

I guess the point is that identity value should not be `Traits<InitT>::Max()` nor `std::numeric_limits<T>::max()` but should be INF. (correct me if I'm wrong @asi1024)

**Comment by leofang** [comment]
Thanks @kmaehashi @asi1024. I missed that you were referring to the reduction identity. It makes sense now.

**Comment by leofang** [comment]
The fix is in: https://github.com/NVIDIA/cub/pull/652. We should start considering the proposal https://github.com/cupy/cupy/issues/6700 to promptly include new Thrust/CUB fixes.

**Comment by betatim** [comment]
I see the same problem with cupy version 13.0.0:
```python
In [3]: cp.min(cp.array([cp.inf, cp.inf]))
Out[3]: array(1.79769313e+308)
```

Is this because the bug has come back, is caused by something else or did the fix never make it in to the release?

Environment info:
```
OS                           : Linux-5.15.0-94-generic-x86_64-with-glibc2.35
Python Version               : 3.12.1
CuPy Version                 : 13.0.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.26.3
SciPy Version                : 1.12.0
Cython Build Version         : 0.29.37
Cython Runtime Version       : 3.0.8
CUDA Root                    : /usr/local/cuda
nvcc PATH                    : /usr/local/cuda/bin/nvcc
CUDA Build Version           : 12000
CUDA Driver Version          : 12020
CUDA Runtime Version         : 12000 (linked to CuPy) / 12010 (locally installed)
cuBLAS Version               : (available)
cuFFT Version                : 11002
cuRAND Version               : 10304
cuSOLVER Version             : (11, 4, 4)
cuSPARSE Version             : (available)
NVRTC Version                : (12, 3)
Thrust Version               : 200001
CUB Build Version            : 200200
Jitify Build Version         : c08b8c6
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Tesla T4
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:3B:00.0
Device 1 Name                : Tesla T4
Device 1 Compute Capability  : 75
Device 1 PCI Bus ID          : 0000:5E:00.0
Device 2 Name                : Tesla T4
Device 2 Compute Capability  : 75
Device 2 PCI Bus ID          : 0000:AF:00.0
Device 3 Name                : Tesla T4
Device 3 Compute Capability  : 75
Device 3 PCI Bus ID          : 0000:D8:00.0
```

**Comment by leofang** [comment]
I suspect it's because the upstream fix was merged after v2.1.0 was cut, but it is still odd to me as we currently pinned at a later commit. Anyway, when I tried to investigate the issue, I hit a new error... https://github.com/NVIDIA/cccl/issues/1373.

**Comment by leofang** [comment]
Sorry, apparently I misread @betatim's example, which involves `min()` not `argmin()`, so it's a new bug. I opened #8180 to track it.

As for the old bug on `argmin()`, I confirm it's fixed upstream and the submodule already included the fix, so let me close this issue.
